File: elasticsearch-java/src/test/java/dev/cerbos/queryplan/elasticsearch/Corpus.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.elasticsearch;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.util.JsonFormat;

import dev.cerbos.api.v1.engine.Engine.PlanResourcesFilter;
import dev.cerbos.api.v1.response.Response.PlanResourcesResponse;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

/**
 * Reads the shared {@code ../conformance/} corpus: the recorded goldens, the PDP tags they were
 * recorded against, and this adapter's ledger. Also holds the one mapping every case translates
 * through. See conformance/README.md, "The harness contract".
 *
 * <p>Every adapter keeps its own copy of this loader; do not extract a shared one (ADR 0007).
 */
final class Corpus {

    private Corpus() {}

    private static final ObjectMapper JSON = new ObjectMapper();

    static Path conformanceDir() {
        return Path.of(System.getProperty("user.dir"), "..", "conformance").normalize();
    }

    private static JsonNode read(Path path) {
        try {
            return JSON.readTree(path.toFile());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    // -- conformance/pdp-versions.json and conformance/golden/ ----------------------------------

    /** The PDP tags the goldens were recorded against: current first, then previous. */
    static final List<String> PDP_TAGS = Stream.of("current", "previous")
            .map(which -> read(conformanceDir().resolve("pdp-versions.json")).get(which).get("tag").asText())
            .toList();

    static final String CURRENT_TAG = PDP_TAGS.get(0);

    /**
     * The planner folds {@code now() - duration("24h")} into a literal, which the goldens record
     * as this placeholder.
     */
    static final String NOW_PLACEHOLDER = "__NOW_MINUS_24H__";

    /**
     * The real instant that goes back in, at the nanosecond precision the planner emits. The
     * fractional digits are fixed, so a run never lands on a whole millisecond by chance.
     */
    static final String NOW_MINUS_24H = Instant.now().minus(24, ChronoUnit.HOURS)
            .truncatedTo(ChronoUnit.SECONDS).plusNanos(123_456_789).toString();

    /** One golden file: a recorded plan and the seed ids {@code check()} allowed. */
    record Golden(String id, String tier, PlanResourcesResponse plan, List<String> allowed,
                  JsonNode plannerDivergence) {

        /** A planner bug for this tag: the plan and {@code check()} disagree, so no adapter can pass. */
        boolean skipped(String tag) {
            return plannerDivergence != null && !plannerDivergence.isNull()
                    && (!plannerDivergence.has("pdp") || strings(plannerDivergence.get("pdp")).contains(tag));
        }
    }

    /** Every golden recorded against one PDP tag, sorted by case id. */
    static List<Golden> goldens(String tag) {
        Path directory = conformanceDir().resolve("golden").resolve(tag);
        try (Stream<Path> files = Files.walk(directory)) {
            return files.filter(path -> path.toString().endsWith(".json")).sorted()
                    .map(path -> golden(path, NOW_MINUS_24H)).toList();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    /** The recorded plan for one case under the current PDP, with the placeholder filled. */
    static PlanResourcesResponse plan(String caseId) {
        return plan(caseId, NOW_MINUS_24H);
    }

    static PlanResourcesResponse plan(String caseId, String nowMinus24h) {
        return golden(conformanceDir().resolve("golden").resolve(CURRENT_TAG)
                .resolve(caseId + ".json"), nowMinus24h).plan();
    }

    private static Golden golden(Path path, String nowMinus24h) {
        JsonNode golden;
        try {
            golden = JSON.readTree(Files.readString(path).replace(NOW_PLACEHOLDER, nowMinus24h));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        // The plan is the PlanResources `filter`, in protobuf's JSON mapping.
        PlanResourcesFilter.Builder filter = PlanResourcesFilter.newBuilder();
        try {
            JsonFormat.parser().merge(golden.get("plan").toString(), filter);
        } catch (InvalidProtocolBufferException e) {
            throw new IllegalStateException(path + " carries a plan protobuf cannot decode", e);
        }
        return new Golden(golden.get("id").asText(), golden.get("tier").asText(),
                PlanResourcesResponse.newBuilder().setFilter(filter).build(),
                strings(golden.get("allowed")), golden.get("plannerDivergence"));
    }

    private static List<String> strings(JsonNode array) {
        return JSON.convertValue(array, new TypeReference<List<String>>() {});
    }

    // -- conformance-ledger.json ----------------------------------------------------------------

    /** One ledger entry: {@code unsupported} or {@code divergent}, optionally limited to some tags. */
    @JsonIgnoreProperties(ignoreUnknown = true)
    record LedgerEntry(String status, String reason, String issue, List<String> pdp) {}

    static final Map<String, LedgerEntry> LEDGER = JSON.convertValue(
            read(Path.of(System.getProperty("user.dir"), "conformance-ledger.json")).get("cases"),
            new TypeReference<LinkedHashMap<String, LedgerEntry>>() {});

    /** The ledger entry for a case under one PDP tag, or null. */
    static LedgerEntry ledgerEntry(String caseId, String tag) {
        LedgerEntry entry = LEDGER.get(caseId);
        return entry != null && (entry.pdp() == null || entry.pdp().contains(tag)) ? entry : null;
    }

    // -- the corpus mapped onto the index --------------------------------------------------------

    /**
     * The corpus field map. Every path is mapped, including ones only refused cases read, so an
     * unmapped field throws {@link UnmappedAttributeException} rather than passing for a refusal.
     */
    static final Map<String, String> FIELD_MAP = Map.ofEntries(
            // The primary key (`request.resource.id`), mapped to the indexed `id` keyword field
            // rather than Elasticsearch's `_id` metadata field.
            Map.entry("request.resource.id", "id"),
            Map.entry("request.resource.attr.aBool", "aBool"),
            Map.entry("request.resource.attr.aString", "aString"),
            Map.entry("request.resource.attr.aNumber", "aNumber"),
            Map.entry("request.resource.attr.aDouble", "aDouble"),
            Map.entry("request.resource.attr.aOptionalString", "aOptionalString"),
            Map.entry("request.resource.attr.createdBy", "createdBy"),
            Map.entry("request.resource.attr.createdAt", "createdAt"),
            Map.entry("request.resource.attr.updatedAt", "updatedAt"),
            Map.entry("request.resource.attr.owner", "owner"),
            // `coOwner` is the explicit-null alias of `scope`, which is omitted when NULL, so the
            // corpus carries the same field under both conventions.
            Map.entry("request.resource.attr.coOwner", "coOwner"),
            Map.entry("request.resource.attr.scope", "scope"),
            Map.entry("request.resource.attr.obj.inner", "obj.inner"),
            Map.entry("request.resource.attr.tags", "tags"),
            Map.entry("request.resource.attr.tagNames", "tagNames"),
            // Flat number and boolean arrays, like tagNames.
            Map.entry("request.resource.attr.aNumberList", "aNumberList"),
            Map.entry("request.resource.attr.aBoolList", "aBoolList"),
            Map.entry("request.resource.attr.categories", "categories"),
            Map.entry("request.resource.attr.mainCategory.subCategories", "mainCategory.subCategories"),
            Map.entry("request.resource.attr.mainCategory.subNames", "mainCategory.subNames"),
            // The corpus's to-one relation, indexed as plain objects because
            // Elasticsearch has no join.
            Map.entry("request.resource.attr.parent.aBool", "parent.aBool"),
            Map.entry("request.resource.attr.parent.aString", "parent.aString"),
            Map.entry("request.resource.attr.parent.aNumber", "parent.aNumber"),
            Map.entry("request.resource.attr.parent.aOptionalString", "parent.aOptionalString"),
            Map.entry("request.resource.attr.parent.inner.aBool", "parent.inner.aBool"),
            Map.entry("request.resource.attr.parent.inner.aString", "parent.inner.aString"),
            Map.entry("request.resource.attr.parent.inner.aNumber", "parent.inner.aNumber"),
            Map.entry("request.resource.attr.parent.inner.aOptionalString", "parent.inner.aOptionalString"));

    /**
     * Attributes the corpus sends to {@code check()} as explicit nulls ({@code resources.json}
     * shows {@code "owner": null}). Every other attribute follows the omitted convention: a NULL
     * column is a missing attribute, which is this adapter's default. Elasticsearch does not index
     * a JSON null, so declaring these makes the adapter refuse instead of answering.
     */
    static final Set<String> EXPLICIT_NULL_ATTRIBUTES = Set.of(
            "request.resource.attr.owner", "request.resource.attr.coOwner");

    /** The field paths the corpus index maps as Elasticsearch {@code nested} documents. */
    static final Set<String> NESTED_PATHS = Set.of(
            "tags", "mainCategory.subCategories",
            "categories", "categories.subCategories", "categories.subCategories.labels");

    /**
     * The field paths the corpus index maps as flat arrays of scalars. The adapter cannot tell
     * {@code size(aString)} from {@code size(tagNames)} by itself, so it refuses {@code size()} on
     * a field declared neither here nor in {@link #NESTED_PATHS}.
     */
    static final Set<String> COLLECTION_FIELDS = Set.of("tagNames", "aNumberList", "aBoolList");

    /**
     * The CEL scalar type of every scalar field in the corpus index, keyed by Elasticsearch field
     * name. The adapter refuses a comparison against an undeclared field, because Elasticsearch
     * coerces a query term to the field's mapped type where CEL's cross-type equality is false.
     */
    static final Map<String, ElasticsearchQueryPlanAdapter.ScalarType> SCALAR_TYPES = scalarTypes();

    private static Map<String, ElasticsearchQueryPlanAdapter.ScalarType> scalarTypes() {
        ElasticsearchQueryPlanAdapter.ScalarType string = ElasticsearchQueryPlanAdapter.ScalarType.STRING;
        ElasticsearchQueryPlanAdapter.ScalarType number = ElasticsearchQueryPlanAdapter.ScalarType.NUMBER;
        ElasticsearchQueryPlanAdapter.ScalarType bool = ElasticsearchQueryPlanAdapter.ScalarType.BOOLEAN;
        ElasticsearchQueryPlanAdapter.ScalarType timestamp = ElasticsearchQueryPlanAdapter.ScalarType.TIMESTAMP;
        Map<String, ElasticsearchQueryPlanAdapter.ScalarType> types = new LinkedHashMap<>();
        types.put("id", string);
        types.put("aString", string);
        types.put("aOptionalString", string);
        types.put("aNumber", number);
        types.put("aDouble", number);
        types.put("aBool", bool);
        types.put("createdAt", timestamp);
        types.put("updatedAt", timestamp);
        types.put("createdBy", string);
        types.put("scope", string);
        types.put("owner", string);
        types.put("coOwner", string);
        types.put("obj.inner", string);
        types.put("tagNames", string);
        types.put("aNumberList", number);
        types.put("aBoolList", bool);
        types.put("mainCategory.subNames", string);
        types.put("mainCategory.subCategories.name", string);
        types.put("tags.id", string);
        types.put("tags.name", string);
        types.put("categories.name", string);
        types.put("categories.subCategories.name", string);
        types.put("categories.subCategories.labels.name", string);
        for (String level : List.of("parent.", "parent.inner.")) {
            types.put(level + "aBool", bool);
            types.put(level + "aString", string);
            types.put(level + "aNumber", number);
            types.put(level + "aOptionalString", string);
        }
        return Map.copyOf(types);
    }

    /** The one mapping every case translates through. */
    static final ElasticsearchQueryPlanAdapter.Options OPTIONS =
            ElasticsearchQueryPlanAdapter.Options.of(FIELD_MAP)
                    .withNestedPaths(NESTED_PATHS)
                    .withScalarTypes(SCALAR_TYPES)
                    .withCollectionFields(COLLECTION_FIELDS)
                    .withExplicitNullAttributes(EXPLICIT_NULL_ATTRIBUTES);

    /** Translates one case's current-PDP plan through {@link #OPTIONS}. */
    static ElasticsearchQueryPlanAdapter.Result translate(String caseId) {
        return ElasticsearchQueryPlanAdapter.toElasticsearchQuery(plan(caseId), OPTIONS);
    }
}

File: prisma/src/corpus.ts
import * as fs from "fs";
import * as path from "path";

import {
  PlanExpression,
  PlanExpressionValue,
  PlanExpressionVariable,
} from "@cerbos/core";
import type {
  PlanExpressionOperand,
  PlanResourcesResponse,
  Value,
} from "@cerbos/core";

import { PlanKind, MapperConfig } from ".";

/**
 * What both of this adapter's suites read from the shared `../conformance/` corpus: the recorded
 * golden plans and the one mapper every case is translated through. `adversarial.test.ts` replays
 * the goldens against a real store; `translator.test.ts` uses the same mapper for the caller-option
 * tests the corpus cannot vary.
 *
 * Duplicated across adapters on purpose — adapters share data, not code (ADR 0007).
 * Test-only: excluded from `tsc --build` by `tsconfig.json`, so nothing here reaches `lib/`.
 */

export const CONFORMANCE_DIR = path.join(__dirname, "..", "..", "conformance");

export function readCorpusJson(file: string): unknown {
  return JSON.parse(fs.readFileSync(path.join(CONFORMANCE_DIR, file), "utf8"));
}

/** The Prisma model the corpus's resource kind maps onto. */
export const MODEL = "AdversarialResource";

// -- the golden files ----------------------------------------------------------------------------

interface WireOperand {
  expression?: { operator: string; operands: WireOperand[] };
  variable?: string;
  value?: unknown;
}

export interface Golden {
  id: string;
  pdp: string;
  tier: "core" | "extended" | "adversarial";
  plan: { kind: string; condition?: WireOperand };
  allowed: string[];
  plannerDivergence: { pdp?: string[]; issue: string; reason?: string } | null;
}

/** The PDP tags the goldens were recorded against: current first, then previous. */
export function pdpTags(): string[] {
  const versions = readCorpusJson("pdp-versions.json") as Record<
    "current" | "previous",
    { tag: string }
  >;
  return [versions.current.tag, versions.previous.tag];
}

/** Every golden file recorded against `tag`, sorted by case id. */
export function readGoldens(tag: string): Golden[] {
  const root = path.join(CONFORMANCE_DIR, "golden", tag);
  const files = (fs.readdirSync(root, { recursive: true }) as string[])
    .filter((file) => file.endsWith(".json"))
    .map((file) => JSON.parse(fs.readFileSync(path.join(root, file), "utf8")) as Golden);
  return files.sort((a, b) => a.id.localeCompare(b.id));
}

export function readGolden(tag: string, id: string): Golden {
  return JSON.parse(
    fs.readFileSync(path.join(CONFORMANCE_DIR, "golden", tag, `${id}.json`), "utf8")
  ) as Golden;
}

/**
 * The instant substituted for `__NOW_MINUS_24H__`, the literal the planner folds
 * `now() - duration("24h")` into. The PDP folds its clock at nanosecond precision, so the
 * substitute carries sub-millisecond digits too: a tidy millisecond instant would translate here
 * while the same case refuses in production (`Timestamp value exceeds millisecond precision`).
 */
export function nowMinus24h(): string {
  const ms = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString();
  return `${ms.slice(0, -1)}456789Z`;
}

function operandFromWire(node: WireOperand, now: string): PlanExpressionOperand {
  if (node.expression) {
    return new PlanExpression(
      node.expression.operator,
      node.expression.operands.map((child) => operandFromWire(child, now))
    );
  }
  if (node.variable !== undefined) {
    return new PlanExpressionVariable(node.variable);
  }
  // The golden is JSON the PDP produced, so its leaves are already the shapes `Value` admits.
  return new PlanExpressionValue(
    (node.value === "__NOW_MINUS_24H__" ? now : node.value) as Value
  );
}

/** A golden's plan decoded the way `@cerbos/http` decodes a PlanResources response. */
export function planOf(
  golden: Golden,
  now: string = nowMinus24h()
): PlanResourcesResponse {
  const base = { cerbosCallId: "", requestId: "", validationErrors: [], metadata: undefined };
  const { kind, condition } = golden.plan;
  if (kind === PlanKind.CONDITIONAL && condition) {
    return { ...base, kind: PlanKind.CONDITIONAL, condition: operandFromWire(condition, now) };
  }
  if (kind === PlanKind.ALWAYS_ALLOWED || kind === PlanKind.ALWAYS_DENIED) {
    return { ...base, kind };
  }
  throw new Error(`${golden.id}: unrecognised plan ${JSON.stringify(golden.plan)}`);
}

// -- the mapper ----------------------------------------------------------------------------------

export const MAPPER: Record<string, MapperConfig> = {
  // The primary key, reached as `request.resource.id` rather than through `attr` (the
  // `identifier/*` cases). It is a mapping like any other here, which is the point: an adapter that resolves
  // references by stripping a `request.resource.attr.` prefix never sees this name.
  "request.resource.id": { field: "id", valueType: "string" },
  // resources.json OMITS aBool, aString and aNumber on the row where the column is NULL (j3, j1,
  // j2), so each follows the omitted convention like aOptionalString below.
  "request.resource.attr.aBool": {
    field: "aBool",
    valueType: "boolean",
    nullable: true,
    nullAttributeRepresentation: "omitted",
  },
  "request.resource.attr.aString": {
    field: "aString",
    valueType: "string",
    nullable: true,
    nullAttributeRepresentation: "omitted",
  },
  "request.resource.attr.aNumber": {
    field: "aNumber",
    valueType: "number",
    nullable: true,
    nullAttributeRepresentation: "omitted",
  },
  "request.resource.attr.aDouble": { field: "aDouble", valueType: "number" },
  // resources.json OMITS aOptionalString when the column is NULL, so `== null` against it is a
  // missing-attribute error the PDP denies, never an IS NULL match.
  "request.resource.attr.aOptionalString": {
    field: "aOptionalString",
    valueType: "string",
    nullable: true,
    nullAttributeRepresentation: "omitted",
  },
  "request.resource.attr.createdBy": { field: "createdBy", valueType: "string" },
  "request.resource.attr.scope": { field: "scope", valueType: "string", nullable: true },
  "request.resource.attr.createdAt": {
    field: "createdAt",
    valueType: "dateTime",
  },
  "request.resource.attr.updatedAt": {
    field: "updatedAt",
    valueType: "dateTime",
  },
  // `owner` and `coOwner` alias columns that `aOptionalString` and `scope` also map, under the
  // OTHER null convention: the oracle sends a real null attribute for them rather than omitting
  // it. Declaring that here is what makes the equality family definite for these two attributes
  // and leaves it untouched for every other mapping (cerbos/query-plan-adapters#308).
  "request.resource.attr.owner": {
    field: "aOptionalString",
    valueType: "string",
    nullable: true,
    nullAttributeRepresentation: "explicit",
  },
  "request.resource.attr.coOwner": {
    field: "scope",
    valueType: "string",
    nullable: true,
    nullAttributeRepresentation: "explicit",
  },
  // obj.inner is not a real nested column — mirrors aString, same trick the spring-data reference
  // harness uses for the comparison/equals/nested-map-member case. `parent.inner` below is the
  // opposite: a real two-level join. The two are kept side by side on purpose.
  "request.resource.attr.obj.inner": {
    field: "aString",
    valueType: "string",
    nullable: true,
    nullAttributeRepresentation: "omitted",
  },
  // The corpus's one REAL to-one chain (the `relation/*` cases). `type: "one"` is what makes the
  // adapter emit `is:` rather than `some:`, and `is:` on an optional relation is what requires
  // the hop to exist — the absent-parent guard the negated shapes discriminate. `inner` nests
  // the same declaration one level further out; two levels is where alias scoping breaks.
  "request.resource.attr.parent": {
    relation: {
      name: "parent",
      type: "one",
      model: "AdversarialParent",
      fields: {
        aBool: { field: "aBool", valueType: "boolean", nullable: true },
        aString: { field: "aString", valueType: "string", nullable: true },
        aNumber: { field: "aNumber", valueType: "number", nullable: true },
        aOptionalString: { field: "aOptionalString", valueType: "string", nullable: true  },
        inner: {
          relation: {
            name: "inner",
            type: "one",
            model: "AdversarialInner",
            fields: {
              aBool: { field: "aBool", valueType: "boolean", nullable: true },
              aString: { field: "aString", valueType: "string", nullable: true },
              aNumber: { field: "aNumber", valueType: "number", nullable: true },
              aOptionalString: { field: "aOptionalString", valueType: "string", nullable: true  },
            },
          },
        },
      },
    },
  },
  "request.resource.attr.tags": {
    relation: {
      name: "tags",
      type: "many",
      // Model name enables field-to-field comparisons between tag columns. `name` is NULLable
      // in the schema, so it keeps the adapter's three-valued-logic guards for collection
      // macros over elements whose name column is NULL (a missing attribute — a CEL error,
      // hence deny — on the check side). Those guards are the default; every REQUIRED element
      // column below says `nullable: false`, because Prisma rejects `{ name: null }` on one.
      model: "AdversarialTag",
      fields: {
        id: { field: "tagId", nullable: false },
        name: { field: "name", valueType: "string", nullable: true  },
      },
    },
  },
  "request.resource.attr.tagNames": {
    relation: {
      name: "tags",
      type: "many",
      field: "name",
      fields: { name: { field: "name", valueType: "string", nullable: true  } },
    },
  },
  // The two homogeneous scalar lists, one owned row per element, projected the way `tagNames`
  // is. `valueType` sits on the list mapping itself because that is where a membership test reads
  // it: a literal of another type (`"2" in aNumberList`) is refused rather than bound, since CEL's
  // heterogeneous equality answers it false and a store that coerces would not. The element column
  // is nullable because a6 and a4 hold a null element, which is a VALUE in CEL.
  "request.resource.attr.aNumberList": {
    valueType: "number",
    relation: {
      name: "numberList",
      type: "many",
      field: "value",
      fields: { value: { field: "value", valueType: "number", nullable: true } },
    },
  },
  "request.resource.attr.aBoolList": {
    valueType: "boolean",
    relation: {
      name: "boolList",
      type: "many",
      field: "value",
      fields: { value: { field: "value", valueType: "boolean", nullable: true } },
    },
  },
  "request.resource.attr.categories": {
    relation: {
      name: "categories",
      type: "many",
      fields: {
        name: { field: "name", valueType: "string", nullable: false },
        subCategories: {
          relation: {
            name: "subCategories",
            type: "many",
            fields: {
              name: { field: "name", valueType: "string", nullable: false },
              labels: {
                relation: {
                  name: "labels",
                  type: "many",
                  fields: {
                    name: { field: "name", valueType: "string", nullable: true  },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
  // The multi-hop chain (the `relation/*/*-to-one-chain` cases): mainCategory mirrors the SAME
  // categories/subCategories relation as a single-object dotted chain on the check side (every
  // seed holds at most one category), pinning that the adapter joins through every intermediate
  // hop, never off the root.
  "request.resource.attr.mainCategory": {
    relation: {
      name: "categories",
      type: "many",
      fields: {
        name: { field: "name", valueType: "string", nullable: false },
        subCategories: {
          relation: {
            name: "subCategories",
            type: "many",
            fields: { name: { field: "name", valueType: "string", nullable: false } },
          },
        },
        // subNames: the same 2-hop chain but with a bare `field`, so plain `in` membership
        // compares the flattened tail's name column directly.
        subNames: {
          relation: {
            name: "subCategories",
            type: "many",
            field: "name",
          },
        },
      },
    },
  },
};
